sdk/nodejs/index.ts

```diff
@@ -1,8 +1,9 @@
-import Client from "./api/client.gen.js"
+import Client, { Container } from "./api/client.gen.js"
```

Comment:
This makes sense, but my IDE is still complaining about the Container type.
Client and connect are both available however, and I can jump to either definition.

Reply:
Sorry, I forgot to execute `yarn watch` in the `dagger/sdk/nodejs` directory.

For me, `export { default as Client } from "./api/client.gen"` works, but I'm afraid it might break backwards compatibility for existing NodeJS pipelines as there's no default export anymore.
With that implementation, I can consume `Client` as a named export only.

In order to keep a default export, I have to do `export { default } from "./api/client.gen"
` in `dagger/sdk/nodejs/index.ts`.

With `export * from "./api/client.gen.js"` in `dagger/sdk/nodejs/index.ts` I have access to all the `./api/client.gen.js` types.

These are both files:

```ts
// dagger/sdk/nodejs/index.ts

export { gql } from "graphql-tag"
export { GraphQLClient } from "graphql-request"

export { connect, ConnectOpts, CallbackFct } from "./connect.js"

export { default } from "./api/client.gen"
export * from "./api/client.gen.js"
```
```ts
// pipeline.ts

import Client, { connect, Container } from "@dagger.io/dagger"
// create Dagger client
connect(async (client: Client) => {
    // setup container and
    // define environment variables
    const ctr = client
        .container()
        .from("alpine")
        .with(envVariables({
            ENV_VAR_1: "VALUE 1",
            ENV_VAR_2: "VALUE 2",
            ENV_VAR_3: "VALUE 3",
        }))
        .withExec(["env"])

    // print environment variables
    console.log(await ctr.stdout())
}, { LogOutput: process.stderr })


function envVariables(envs: Record<string, string>) {
    return (c: Container): Container => {
        Object.entries(envs).forEach(([key, value]) => {
            c = c.withEnvVariable(key, value)
        })
        return c
    }
}
```

Reply:
> Sorry, I forgot to execute `yarn watch` in the `dagger/sdk/nodejs` directory.
> 
> For me, `export { default as Client } from "./api/client.gen"` works, but I'm afraid it might break backwards compatibility for existing NodeJS pipelines as there's no default export anymore. With that implementation, I can consume `Client` as a named export only.
> 
> In order to keep a default export, I have to do `export { default } from "./api/client.gen" ` in `dagger/sdk/nodejs/index.ts`.
> 
> With `export * from "./api/client.gen.js"` in `dagger/sdk/nodejs/index.ts` I have access to all the `./api/client.gen.js` types.
> 
> These are both files:
> 
> ```ts
> // dagger/sdk/nodejs/index.ts
> 
> export { gql } from "graphql-tag"
> export { GraphQLClient } from "graphql-request"
> 
> export { connect, ConnectOpts, CallbackFct } from "./connect.js"
> 
> export { default } from "./api/client.gen"
> export * from "./api/client.gen.js"
> ```
> 
> ```ts
> // pipeline.ts
> 
> import Client, { connect, Container } from "@dagger.io/dagger"
> // create Dagger client
> connect(async (client: Client) => {
>     // setup container and
>     // define environment variables
>     const ctr = client
>         .container()
>         .from("alpine")
>         .with(envVariables({
>             ENV_VAR_1: "VALUE 1",
>             ENV_VAR_2: "VALUE 2",
>             ENV_VAR_3: "VALUE 3",
>         }))
>         .withExec(["env"])
> 
>     // print environment variables
>     console.log(await ctr.stdout())
> }, { LogOutput: process.stderr })
> 
> 
> function envVariables(envs: Record<string, string>) {
>     return (c: Container): Container => {
>         Object.entries(envs).forEach(([key, value]) => {
>             c = c.withEnvVariable(key, value)
>         })
>         return c
>     }
> }
> ```

Yes in order to try that, you need to yarn build (not necessarly yarn watch) to update the SDK in your `dagger` directory.
Did you restart typescript on your vscode ? 
<img width="622" alt="image" src="https://github.com/dagger/dagger/assets/1186424/8a7ebe9d-9795-4a1c-b973-4586e9a00533">

If it works in your side @crjm I think we should ship it.
@helderco Like I've said on the description, we should ship that first to fix PR #5139 and then we can create a following PR for everything related to the exports. Thanks

Reply:
> Why not:
> 
> ```ts
> export { default as Client } from "./api/client.gen"
> export * from './api/client.gen';
> ```
> 
> (and remove `export { Container }`)

Actually we just have to do:
```ts
export * from "./api/client.gen.js"
```
The line export { default as Client } from "./api/client.gen" is not required at this point. The default export, Client, is maintained and all other exports are accessible.

Reply:
Yes, I know. I put two things in there. I think even `Client` should be a named export, but that can be left for a follow up. For now it's just better to `export *` rather than just `{ Container }`.

Reply:
Sorry, I meant to keep the default export, but mistyped because I was thinking it would be better for it to be a named export, but that's for another PR.

What I meant was:

```diff
- import Client, { Container } from "./api/client.gen.js"
- export default Client
- export { Container }
+ export { default } from "./api/client.gen"
+ export * from './api/client.gen';
```

I meant this syntax as a simplification for re-exporting, but the important thing here is to re-export all named exports from `client.gen`, not just `Container`. We shouldn't require another PR for adding the other types as it's a simple change.

Reply:
> Sorry, **I meant to keep the default export**, but mistyped because I was thinking it would be better for it to be a named export, but that's for another PR.
> 
> What I meant was:
> 
> ```diff
> - import Client, { Container } from "./api/client.gen.js"
> - export default Client
> - export { Container }
> + export { default } from "./api/client.gen"
> + export * from './api/client.gen';
> ```
> 
> I meant this syntax as a simplification for re-exporting, which should preserve the default export, but the important thing here is to **re-export all named exports** from `client.gen`, not just `Container`. We shouldn't require another PR for adding the other types as it's a simple change.

The first line is still unecessary as `export * from './api/client.gen` with preserve the default export.

Reply:
I've just pushed the fix cc @helderco

Reply:
@slumbering I don't think `export *` preserves the default export.

> There is also `export * from "mod"` [...]. This re-exports all **named exports** from `mod` as the named exports of the current module, but the default export of `mod` is not re-exported.

\- [Re-exporting / Aggregating](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export#re-exporting_aggregating)

Reply:
> I don't think export * preserves the default export.

Can confirm that `export { default } from "./api/client.gen"` is necessary, otherwise the package doesn't have a default export. If I remove that line I get:
```console
Module '"~/dagger/sdk/nodejs/dist/index"' has no default export.ts(1192)
index.d.ts(1, 1): 'export *' does not re-export a default.
```

Reply:
Following @helderco's link + @crjm's consensus. It seems that merging with this implementation:
```nodejs
export { default } from "./api/client.gen"
export * from './api/client.gen';
```
is the best next step forward

Reply:
> @slumbering I don't think `export *` preserves the default export.
> 
> > There is also `export * from "mod"` [...]. This re-exports all **named exports** from `mod` as the named exports of the current module, but the default export of `mod` is not re-exported.
> 
> - [Re-exporting / Aggregating](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export#re-exporting_aggregating)

Sorry my bad. It was working on my machine when I tried it.... maybe I forgot to rebuild the SDK or something else.